### DISCUSSION TITLE
$refsArray helper

### DISCUSSION THREAD

**Discussion by KevinBatdorf** [alpine-collective/alpine-magic-helpers#48] [Ideas]
Pretty simple actually, but how's the naming here?

```js
Alpine.addMagicProperty('refsArray', el => {
  return Array.from(el.querySelectorAll('[x-ref]'))
})
```

https://codepen.io/KevinBatdorf/pen/dypbMra

**Comment by HugoDF** [comment]
`allRefs`? `refList`?

**Comment by KevinBatdorf** [comment]
You think we could extend the default behavior even and have the best of both?

**Reply by KevinBatdorf** [reply]
I'd probably say not to override default behavior because it could lead to people reading code and not understanding it, and just making things confusing for new people. But I'm not too confident on that opinion, so would be interested what anyone else thinks.

**Reply by KevinBatdorf** [reply]
Here's one way that could work:

```html
<template x-for="ref in Object.values($refs)">
  <p x-text="ref.textContent"></p>
</template>
<p x-text="$refs.foo.textContent"></p>
```

```js
Alpine.addMagicProperty('refs', el => {
  let refs = []
  Array.from(el.querySelectorAll('[x-ref]')).forEach(ref => {
    refs[ref.getAttribute('x-ref')] = ref
  })
  return refs
})
```

https://codepen.io/KevinBatdorf/pen/8c62b8d945fb1c909c3b6e4ab43e6477

**Reply by HugoDF** [reply]
`let refs = {}` but yeah

**Reply by HugoDF** [reply]
maybe we could use the name `$magicRefs` which works similarly to `$refs` but can be enumerated?

**Reply by HugoDF** [reply]
Oh dear, JS allows this 😄 (Array is an Object something something) usable with `x-for="ref in $refs"` _and_ backwards compatible

```js
Alpine.addMagicProperty('refs', el => {
  let refs = []
  Array.from(el.querySelectorAll('[x-ref]')).forEach(ref => {
	  refs.push(ref);
	  refs[ref.getAttribute('x-ref')] = ref
  })
  return refs
})
```

**Reply by KevinBatdorf** [reply]
> let refs = {} but yeah

True. Updated the demo

I like `$magicRefs`

**Reply by HugoDF** [reply]
I think a different name just to make it obvious it's not built-in (& heopfully avoid the issues on the main repo) but keep similar functionality to smooth out the learning curve.

**Reply by KevinBatdorf** [reply]
> Oh dear, JS allows this 😄 (Array is an Object something something) usable with x-for="ref in $refs" and backwards compatible

Oh wow, feels so wrong though 😈

**Reply by KevinBatdorf** [reply]
Yeah I think even if we don't document it people will expect `$magicRefs.something` to work. But we can get away with having them use `Object.values($magicRefs)`. You think?

**Reply by HugoDF** [reply]
> Oh wow, feels so wrong though 😈

Oh yeah, but then again why don't we use JS quirks for good 😂 

> Yeah I think even if we don't document it people will expect $magicRefs.something to work. But we can get away with having them use Object.values($magicRefs). You think?

yeah I think that's more reasonable, works like `$refs` except it's enumerable so Object.keys/values/entries will work

**Reply by ryangjchandler** [reply]
How about keeping the nature of the built-in `$refs` and instead extending it with a `.all()` method, that returns the custom stuff?

**Reply by KevinBatdorf** [reply]
Do you think it's a good idea to hijack and override `$refs`? I'm not against it I'm just unsure. It would also break if Alpine changes `$refs` someday.

Would this be better as a PR to Alpine?

**Reply by SimoTod** [reply]
Using the same name may result in people opening issues on the main repo. I know it's obvious to us but a lot of users are just cut-paster, they don't read the readme (or they do but english is not their main language and they misunderstand it).
That being said, this implementation of refs is more inline with what the Alpine community expect and want so it's hard to say.
I think Caleb wants to remove x-ref in v3 (when and if it happens) so putting out a helper with the same name but a "better" behaviour may push it towards removing $refs from the core which may be what he wants anyway.

**Reply by KevinBatdorf** [reply]
> a lot of users are just cut-paster, they don't read the readme

My initial thought too. 

We could PR Alpine to remove `$refs` and remove it from the docs, and add a note to instruct people to come here. Kind of a bold thing to do though 😏

**Reply by SimoTod** [reply]
$betterRefs 😜
The PR to Alpine is probably too much. It should go through a private channel first to make sure Caleb is okay with that so it doesn't look like we are stepping on his feet.
So we have these possible paths:
- we use a different name (exact name TBC)
- we use $refs, extending the current implementation, we make clear in the readme what is going on but we don't touch Alpine
- we ask Caleb, if he agrees we implement the above and after a transition period we PR something to remove refs from core.
- we extend refs to support all() 

I think I'm going to sleep on it.

**Reply by SimoTod** [reply]
Should we also support references using the same name?

```
<div x-ref="foo">one</div>
<div x-ref="foo">two</div>
```
`$getRefs.foo` would return `[<div x-ref="foo">one</div>, <div x-ref="foo">two</div>]`

**Reply by SimoTod** [reply]
P.s. Although it's tempting I think we shouldn't use the same name to avoid confusion and conflicts in the future

**Reply by KevinBatdorf** [reply]
> Should we also support references using the same name?

I think we should support that, yes. 

For the name, we can start with anything for now and if it ends up being more popular than `$refs` we can just alias it later on, maybe even alias it via a config flag or something.

We could even just do the full version,`$references`
